addons/binding/org.openhab.binding.nest/src/main/java/org/openhab/binding/nest/handler/NestBaseHandler.java

```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.nest.handler;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.smarthome.core.library.types.DateTimeType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.openhab.binding.nest.NestBindingConstants;
+import org.openhab.binding.nest.internal.NestDeviceDataListener;
+import org.openhab.binding.nest.internal.NestUpdateRequest;
+
+import java.util.Calendar;
+import java.util.Date;
+import java.util.Optional;
+import java.util.TimeZone;
+
+/**
+ * Deals with the structures on the nest api, turning them into a thing in openhab.
+ *
+ * @author David Bennett - initial contribution
+ * @author Martin van Wingerden - splitted of NestBaseHandler
+ */
+abstract class NestBaseHandler extends BaseThingHandler implements NestDeviceDataListener {
+    NestBaseHandler(@NonNull Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        getNestBridgeHandler()
+                .map(b -> b.addDeviceDataListener(NestBaseHandler.this));
```

Comment:
I run the binding with debug logging enabled after a while I see that things are updated more than once after data has been updated from Nest. When you check the call hierarchy of this initialize() method I think this explains. Because of this I think the listener is added several times to its NestBridgeHandler.

Reply:
You are right, I made a mistake here. All handlers should only process updates for there own things, now all smoke sensors handlers process all data of all smoke sensors.

Reply:
The issue with unconditionally adding a listener in this method is that it also gets called by `BaseThingHandler.thingUpdated(..)` and `BaseThingHandler.handleConfigurationUpdate(..)`.

The issue you describe seems to be happening in the `NestBridgeHandler.broadcastDevices(..)` and `broadcastStructure()` methods. In those methods each device/structure gets an update of all devices/structures. Which results in all devices/structures getting multiple updates and ending up with the same status (of the last device of that type).

Reply:
Yes but I would hope the first part is resolved by unregistering it in the dispose. AFAIK the call is dispose thingUpdated (re)initialize